Sources/NIOHPACK/IndexedHeaderTable.swift

```diff
@@ -0,0 +1,150 @@
+        if index < self.staticTable.count {
+            let entry = self.staticTable[index]    // static table is already 1-based
+            views = (self.staticTable.view(of: entry.name), self.staticTable.view(of: entry.value))
+        } else if index - self.staticTable.count < self.dynamicTable.count {
+            let entry = self.dynamicTable[index - self.staticTable.count]
+            views = (self.dynamicTable.view(of: entry.name), self.dynamicTable.view(of: entry.value))
+        } else {
+            throw NIOHPACKErrors.InvalidHeaderIndex(suppliedIndex: index, availableIndex: self.staticTable.count + self.dynamicTable.count - 1)
+        }
+        
+        return (String(decoding: views.0, as: UTF8.self), String(decoding: views.1, as: UTF8.self))
+    }
+    
+    /// Obtains the header key/value pair at the given index within the table as sequences of
+    /// raw bytes.
+    ///
+    /// - note: Per RFC 7541, this uses a *1-based* index.
+    /// - Parameter index: The index to query.
+    /// - Returns: A tuple containing the name and value of the stored header.
+    /// - Throws: `NIOHPACKErrors.InvalidHeaderIndex` if the supplied index was invalid.
+    public func headerViews(at index: Int) throws -> (name: RingBufferView, value: RingBufferView) {
+        let result: (RingBufferView, RingBufferView)
+        if index < self.staticTable.count {
+            let entry = self.staticTable[index]
+            result = (self.staticTable.view(of: entry.name), self.staticTable.view(of: entry.value))
+        } else if index - self.staticTable.count < self.dynamicTable.count {
+            let entry = self.dynamicTable[index - self.staticTable.count]
+            result = (self.dynamicTable.view(of: entry.name), self.dynamicTable.view(of: entry.value))
+        } else {
+            throw NIOHPACKErrors.InvalidHeaderIndex(suppliedIndex: index, availableIndex: self.staticTable.count + self.dynamicTable.count - 1)
+        }
+        
+        return result
+    }
+    
+    /// Searches the table to locate an existing header with the given name and value. If
+    /// no item exists that contains a matching value, it will return the index of the first
+    /// item with a matching header name instead, to be encoded as index+value.
+    ///
+    /// - Parameters:
+    ///   - name: The name of the header to locate.
+    ///   - value: The value for which to search.
+    /// - Returns: A tuple containing the index of any located header, and a boolean indicating
+    ///            whether the item at that index also contains a matching value. Returns `nil`
+    ///            if no match could be found.
+    public func firstHeaderMatch(for name: String, value: String?) -> (index: Int, matchesValue: Bool)? {
+        return self.firstHeaderMatch(for: name.utf8, value: value?.utf8)
+    }
+    
+    func firstHeaderMatch<C : Collection>(for name: C, value: C?) -> (index: Int, matchesValue: Bool)? where C.Element == UInt8 {
```

Comment:
This function can also do with an optimisation for the case where `value` is `nil`, as noted previously.